lebedigital/openbis/expstep.py

```diff
@@ -0,0 +1,1464 @@
+        if not 'files' in props:
+            raise KeyError(f'files not specified in props')
+        if not 'data_type' in props:
+            raise KeyError(f'data_type not specified in props')
+
+        files = props.pop('files')
+        data_type = props.pop('data_type')
+
+        # If a dataset with the same name was found in the datastore that dataset will be returned and none will be uploaded
+        if test:
+            name = props['$name']
+            print(
+                f'Dataset(s) with the same name already present in the Datastore.\nTo upload the dataset you must first delete the other dataset with name {name}')
+            ds = o.get_datasets(where={'$name': props['$name']})
+
+        # Uploading the dataset
+        else:
+            ds = o.new_dataset(
+                type=data_type,
+                collection=self.collection,
+                sample=self.identifier,
+                files=files,
+                props=props
+            )
+            ds.save()
+
+        return ds
+
+    def download_datasets(self, o: Openbis, path: str, data_type: str = ''):
+        """Downloads all datasets which are asigned to that sample
+
+        Args:
+            o (Openbis): Currently running openbis instance
+            path (str): Path where the datasets should be saved
+            data_type (str): If specified will only download data sets of that type
+
+        Raises:
+            ValueError: Raises an error when no datasets are found under the sample
+        """
+
+        # If the sample has no datasets an error will be thrown
+        if not len(self.datasets):
+            raise ValueError('No Datasets found under the sample')
+
+        file_plural = 'FILES' if len(self.datasets) > 1 else 'FILE-'
+
+        print(
+            f'----------DOWNLOADING {len(self.datasets)} {file_plural}----------\n')
+        for dataset in self.datasets:
+            # If data_type was specified download only the datastes with that data type
+            if data_type:
+                if dataset.type == data_type:
+                    print(f'Downloading dataset {dataset.code}')
+                    print(f'Files: {dataset.file_list}\n')
+                    dataset.download(
+                        destination=path,
+                        create_default_folders=False,
+                        wait_until_finished=False,
+                    )
+
+            # If data_type was NOT specfied download all datasets
+            else:
+                print(f'Downloading dataset {dataset.code}')
+                print(f'Files: {dataset.file_list}\n')
+                dataset.download(
+                    destination=path,
+                    create_default_folders=False,
+                    wait_until_finished=False,
+                )
+
+        print('----------DOWNLOAD FINISHED----------')
+
+    @staticmethod
+    def create_sample_type_emodul(o: Openbis, sample_code: str, sample_prefix: str, sample_properties: dict):
```

Comment:
Sample types are blueprints for sample objects that have to be specified before creating a sample object in openBIS. The method is for generating these blueprints from the metadata yaml files created from the metadata extraction step. The user needs to specify the name and prefix of the sample type as they cant be read from the yaml file. 

After creating the sample type you can create an object with the specified metadata. In the 1 file Minimum Working Example it will be the same file but if we created 20 yaml files with the same keys then we would create one sample_type and 20 sample objects.